awscli/customizations/dynamodb/extractor.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from .parser import Parser
+
+
+class AttributeExtractor(object):
+    COMPARATORS = {
+        'eq': '=',
+        'ne': '<>',
+        'lt': '<',
+        'lte': '<=',
+        'gt': '>',
+        'gte': '>=',
+    }
+
+    def __init__(self, parser=None):
+        self._parser = parser
+        if parser is None:
+            self._parser = Parser()
+        self._identifiers = {}
+        self._literals = {}
+        self._index_offset = 0
+
+    def extract(self, expression, index_offset=0):
+        self._identifiers = {}
+        self._literals = {}
+        self._index_offset = index_offset
+        root = self._parser.parse(expression)
+        expression = self._visit(root).strip()
+        return {
+            'expression': expression,
+            'identifiers': self._identifiers,
+            'values': self._literals,
+            'substitution_count': len(self._identifiers) + len(self._literals)
+        }
+
+    def _substitution_index(self):
+        num_substituted = len(self._identifiers) + len(self._literals)
+        return num_substituted + self._index_offset
+
+    def _visit(self, node):
+        method = getattr(self, '_visit_%s' % node['type'])
+        return method(node)
+
+    def _visit_comparator(self, node):
+        left = self._visit(node['children'][0])
+        right = self._visit(node['children'][1])
+        expression = '%s%s %s' % (
```

Comment:
Could you elaborate on why there is no space for the first element but a space for the second element?

Reply:
Talked offline. It is because the left operand always ends with a space.